google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/DynamicFlowControlStats.java

```diff
@@ -72,45 +75,60 @@ boolean setLastAdjustedTimestampMs(long last, long now) {
     private double mean;
     private double weightedCount;
     private AtomicLong lastUpdateTimeInSecond;
+    private long startTimeSecond;
 
     DecayingAverage(double decayConstant) {
       this.decayConstant = decayConstant;
       this.mean = 0.0;
       this.weightedCount = 0.0;
+      this.startTimeSecond = TimeUnit.MILLISECONDS.toSeconds(System.currentTimeMillis());
       this.lastUpdateTimeInSecond = new AtomicLong(0);
     }
 
     synchronized void update(long value, long timestampMs) {
+      updateStartTime(timestampMs);
+
       long now = TimeUnit.MILLISECONDS.toSeconds(timestampMs);
-      Preconditions.checkArgument(
-          now >= lastUpdateTimeInSecond.get(), "can't update an event in the past");
-      if (lastUpdateTimeInSecond.get() == 0) {
+      long elapsed = now - startTimeSecond;
+      double weight = getWeight(elapsed);
+      // Exponential moving average = weightedSum / weightedCount, where
+      // weightedSum(n) = weight(n) * value(n) + weightedSum(n - 1)
+      // weightedCount(n) = weight(n) + weightedCount(n - 1), where weight(n) grows exponentially
+      // over elapsed time.
+      // Using weighted count in case the sum overflows.
+      mean =
+          mean * (weightedCount / (weightedCount + weight))
+              + weight * value / (weightedCount + weight);
+      weightedCount += weight;
+
+      // Set last update time so when we're getting the mean we can calculate the decay based on the
+      // last time the mean was updated.
+      if (now > lastUpdateTimeInSecond.get()) {
         lastUpdateTimeInSecond.set(now);
```

Comment:
why is this still necessary? isnt this done in updateStartTime now?

Reply:
Hmm, so if getMean is called after like 5 minutes of update (before we need to update start time) the average will be decayed?